### ISSUE TITLE
Auto Discover not working?

### ISSUE THREAD

**Issue by devel-q** [elbowz/ha-homie#1]
Hi,
thanks first of all for you work and providing your contribution to HA and Homie!
I was trying out your plugin. Unfortunately I cannot get any devices or entities automatically discovered.

The information is published by mbmd: https://github.com/volkszaehler/mbmd#mqtt-api

My structure for homie is as followed:
homie/sdm1-1/$state disconnected
**homie/sdm1-1/$homie 4.0**
homie/sdm1-1/$name SDM1.1
homie/sdm1-1/$state init
homie/sdm1-1/$implementation MBMD
homie/sdm1-1/$nodes meter
homie/sdm1-1/$implementation (null)
homie/sdm1-1/meter/$name SDM
homie/sdm1-1/meter/$type Eastron SDM630
homie/sdm1-1/meter/importl1/$name L1 Import
homie/sdm1-1/meter/importl1/$unit kWh
...

The values were set via UI, but I also tried it via configuration.yaml. The values are the default ones you have suggested in your howto and given as default values:
base_topic: +
  qos: 2
  discovery: true

Any idea why the devices and entities are not showing up?

**Comment by elbowz** [comment]
Hi @devel-q, thanks for your issue.

This Home Assistant integration is the early stage and I cannot ensure an adequate fast and fully support in this days, so you should have patience.

Some steps and suggestions to go ahead:

1. Enable debug logging and post the output (eg pastebin). So add this to your `configuration.yaml`:
```yaml
logger:
  default: info
  logs:
    custom_components.homie: debug
```

2. I don't know mbmd, I used and tested ha-homie with homie for esp8266 (https://github.com/homieiot/homie-esp8266). 
    Following the Homie convention (https://homieiot.github.io/), you should have something different in the mqtt topic structure:

```
homie / device123 / $homie → 3.0
homie / device123 / $name → My device
homie / device123 / $state → ready
homie / device123 / $nodes → mythermostat

homie / device123 / mythermostat / $name → My thermostat
homie / device123 / mythermostat / $properties → temperature

homie / device123 / mythermostat / temperature → 22 
homie / device123 / mythermostat / temperature / $name → Temperature
homie / device123 / mythermostat / temperature / $unit → °C
homie / device123 / mythermostat / temperature / $datatype → integer
homie / device123 / mythermostat / temperature / $settable → true
```

In your structure missing the `$state` = `ready` the `$datatype` in the property `L1 Import`. Please re-check this information. If the device is not `ready`, discovery cannot work...it also need `$datatype` to understand what kind of entity add to HA.

3. If you wanna do a blinded attempt, try to append "4.0" to the `HOMIE_SUPPORTED_VERSION` list in the file `const.py`:
   `HOMIE_SUPPORTED_VERSION = ["3.0.0", "3.0.1", "4.0.0", "4.0"]`

Ciao!

**Comment by devel-q** [comment]
Hi @elbowz ,

thanks for your quick reply!

I tried to debug a bit:

- $state=ready is available. The information are only too much, this is why I cut them.
- the $datatype exists (float)
- I can provide you a fill file with all the content but its pretty huge. It seems that there are multiple iterations and the list of entities gets longer every time. It then ends the process with $state=ready
- I have added the "4.0" to the list of supported versions. After that I was able to see that new devices through the entire method (last log output was in line 212). But the devices or entities are still not visible in HA.

The only issue I could see (but could also been caused by my temporary added log outputs) was:
`2021-12-13 17:36:27 DEBUG (MainThread) [custom_components.homie] async_device_on_ready ( homie_device = <custom_components.homie.homie.component.HomieDevice object at 0x7ff57d34b5e0> )
2021-12-13 17:36:27 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/config/custom_components/homie/homie/component.py", line 138, in _async_update_topic_dict
    await self._async_on_ready(self)
  File "/config/custom_components/homie/utils.py", line 33, in wrap_f
    return await f(*args, **kwargs)
  File "/config/custom_components/homie/__init__.py", line 221, in async_device_on_ready
    async_create_ha_device(hass, homie_device, entry)
  File "/config/custom_components/homie/mixins.py", line 45, in async_create_ha_device
    mac = device_registry.format_mac(device.t["$mac"])
  File "/usr/src/homeassistant/homeassistant/helpers/device_registry.py", line 121, in format_mac
    if len(to_test) == 17 and to_test.count(":") == 5:
TypeError: object of type 'NoneType' has no len()`

**Comment by elbowz** [comment]
Hi @devel-q,

> $state=ready is available. The information are only too much, this is why I cut them.
the $datatype exists (float)
I can provide you a fill file with all the content but its pretty huge. It seems that there are multiple iterations and the list of entities gets longer every time. It then ends the process with $state=ready

Please provide me the MQTT device topic tree structure (with at least one property) you would like be discovered. Here, an example taken with MQTT Explorer (you should have `homie` instead of `bdiot`:
![Screenshot_20211214_102922-2](https://user-images.githubusercontent.com/1451484/145972791-9cb206aa-4814-43fa-8423-c4a1c5140578.png)

> I have added the "4.0" to the list of supported versions. After that I was able to see that new devices through the entire method (last log output was in line 212). But the devices or entities are still not visible in HA.

I don't understand what you mean. I need the full HA log with the debug enabled (see previous post), posted on pastebin.

> The only issue I could see (but could also been caused by my temporary added log outputs) was:

I think you don't have `homie/device-id/$mac`...for this reason I need the the mqtt topics structure and HA log.

Ciao!

**Comment by devel-q** [comment]
Hi @elbowz,

here an overview with MQTT Explorer:
![image](https://user-images.githubusercontent.com/95985942/145986974-858043dd-3367-41d1-ba13-2f9cd53f5a90.png)
I have opened just one subnode, but the others are similar.
The logfile I have placed here: https://drive.google.com/file/d/14db2kKm6jVRGo8seiKceBbDIG6yAd101

**Comment by elbowz** [comment]
Ok, thanks for your detailed information.

Two things:
* Delete your log from gdrive...I just see there are some sensitive information (local service, but better remove)
* Referring to your mqtt, so don't you have `$mac`, `$localip`, `$stats`, `$fw` or `$implementation`, right?!

~~Try this speedy workaround:~~
* ~~Edit file `mixins.py`~~
* ~~change the line 45 from~~
  ~~`mac = device_registry.format_mac(device.t["$mac"])`~~
  ~~to~~
  ~~`mac = device.id`~~

~~I'm sorry, but I'm quite busy in these days, if you prefer wait some couple of days, I'll create a branch with possible fix...~~

~~edit:~~
* ~~do the same in file `entity_base.py` line 121:~~
  ~~`mac = device_registry.format_mac(self._homie_device.t["$mac"])`~~
  ~~to~~
  ~~`mac = self._homie_device.id`~~
* ~~totally remove: `extra_state_attributes(self)` line 93-116~~

~~Ok, I think should better a test branch :)~~

**Comment by elbowz** [comment]
ok, I have created a dedicated branch, download and try:
https://github.com/elbowz/ha-homie/archive/refs/heads/fix-%231.zip 

...and cross finger

**Comment by devel-q** [comment]
Thank you very much @elbowz  for providing the fix.
After also adding the "4.0" to the supported versions the device did appear. But unfortunately there are no Entities being added.

The logs are not showing any exception here:
`2021-12-14 17:44:09 DEBUG (MainThread) [custom_components.homie] async_discovery_message_received ( mqttmsg = ReceiveMessage(topic='homie/sdm1-1/$homie', payload='4.0', qos=0, retain=0, subscribed_topic='+/+/$homie', timestamp=datetime.datetime(2021, 12, 14, 17, 44, 9, 615312, tzinfo=datetime.timezone.utc)) )`
`2021-12-14 17:44:09 DEBUG (MainThread) [custom_components.homie] async_discovery_message_received ( mqttmsg = ReceiveMessage(topic='homie/sdm1-1/$homie', payload='4.0', qos=0, retain=0, subscribed_topic='+/+/$homie', timestamp=datetime.datetime(2021, 12, 14, 17, 44, 9, 620445, tzinfo=datetime.timezone.utc)) )`
`2021-12-14 17:44:42 DEBUG (MainThread) [custom_components.homie] async_device_on_ready ( homie_device = <custom_components.homie.homie.component.HomieDevice object at 0x7fd3b74fe400> )`

As you said you are currently short in time, please take your time to look into that ;-)

**Comment by elbowz** [comment]
> After also adding the "4.0" to the supported versions the device did appear. But unfortunately there are no Entities being added.

Do you mean `HA > Configuration > Devices` and `HA > Configuration > Entities`, right? 
And you have only a new device?

**Comment by devel-q** [comment]
> Do you mean HA > Configuration > Devices and HA > Configuration > Entities, right?
And you have only a new device?

Yes correct. You can also see it in the integration overview. I have only a device, but in the entity I cannot find any attributes related to that device.

**Comment by elbowz** [comment]
Let's try the updated branch: https://github.com/elbowz/ha-homie/archive/refs/heads/fix-%231.zip

1. remove the old ha-homie from custom component and put this new one
2. keep `custom_components.homie: debug` in configuration.yaml
3. restart HA
4. wait 2-3 mins to allow discovering process
5. save HA log and send me to 24gfy6vk@anonaddy.me (avoiding spread sensible data)
6. make me know the HA version you use [Configuration panel -> Info](https://my.home-assistant.io/redirect/info/) (I just discover it's not in log)

p.s. this version doesn't fix nothing, I have only added more verbosity to debug log

**Comment by devel-q** [comment]
Sorry for my delayed answer. You should have a logfile in your inbox!
I'm using version "core-2021.12.1".

**Comment by elbowz** [comment]
I have updated my HA to 2021.12.3 and the error `AttributeError: module 'homeassistant.components.number' has no attribute 'MODE_AUTO'` appear also in my log. This was fixed thanks to @legovaer on PR #2.

But also without the fix on my side the entities (except the number) was discovered and added...so, I don't know.

I have rebase the branch fix-#1 with the latest PR, added some more verbosity to the debug log.

TL;DR download the latest branch version (https://github.com/elbowz/ha-homie/archive/refs/heads/fix-%231.zip) and redo the point 1-5 of my last post...

If you can, repost the topic tree (device + node + property) you would like to be discovered.

...and sry for all this try&error procedure

**Comment by devel-q** [comment]
You should have again a logfile in your inbox with the latest branch version.

Here is a screenshot with all the attributes:
![image](https://user-images.githubusercontent.com/95985942/146832807-6e9db0c6-9251-4a7a-8a7a-2539bf6a24fa.png)
I'm trying to get those values as entities. For instance "currentl1" (just as an example).

> ...and sry for all this try&error procedure

Don't worry at all. Thank you very much for your support!

**Comment by elbowz** [comment]
Hey @devel-q, have you try at the end the suggested change sent by email?

> you can try to change the line 59 in `homie/component.py`:
> `async def _event_wait(self, name, timeout=10):`
> to
> `async def _event_wait(self, name, timeout=180):`
> 
> and wait more than 3 minutes after HA start (ie. GUI become available)

**Comment by devel-q** [comment]
I've tried that. As well with the code I had and the latest pull from master. Unfortunately same behavior...

**Comment by elbowz** [comment]
> I've tried that. As well with the code I had and the latest pull from master. Unfortunately same behavior...

Could you post the log with the master version with timeout=180? 
Remember to wait at least 4 minutes before copy the log.

I suggest you to insert manually a fake topic in you MQTT (eg. with MQTT explorer) like:

```
homie / device123 / $homie → 3.0
homie / device123 / $name → My device
homie / device123 / $state → ready
homie / device123 / $nodes → mythermostat
homie / device123 / mythermostat / $name → My thermostat
homie / device123 / mythermostat / $properties → temperature
homie / device123 / mythermostat / temperature → 22 
homie / device123 / mythermostat / temperature / $name → Temperature
homie / device123 / mythermostat / temperature / $unit → °C
homie / device123 / mythermostat / temperature / $datatype → integer
homie / device123 / mythermostat / temperature / $settable → true
```

so we can understand if there is something wrong on your Homie devices.

edit: use retained flag

**Comment by elbowz** [comment]
Feel free to open a new one if the issue is not gone...
